### PR TITLE
Basic changes to support poetry 1.2.0, no longer support Poetry 1.1.x

### PR DESCRIPTION
**Pull request by Callek**
This adds support for poetry 1.2.0b2 while dropping 1.1.x support. 

Might still need some work but seems to be working ok

### CODE REVIEW

**Comment by Callek** [comment]
@enpaul this is now ready for review.  Had a CI run over at: https://github.com/Callek/tox-poetry-installer/actions/runs/2638529975

(with a now-dropped commit that added support for CI on my branch for my own fork)

**Comment by Callek** [comment]
@enpaul Updated to support Poetry 1.2.0b3 which was just released.

**Review by fsgnil — APPROVED** [review]
Hi, thanks for fixing this. I'm also looking to use tox-poetry-installer with peotry 1.2.0. I'm not sure if my approval helps. But hopefully this gets merged soon.

**Comment by Mause** [comment]
1.2.0 has now been released, would be great to see this merged to get compatibility with it

**Comment by lihu-zhong** [comment]
My spidey senses tell me we'll see a release this coming Tuesday.

**Comment by zachtaira** [comment]
@Callek thanks for the PR! Much appreciated

**Comment by ctapobep** [comment]
Since Poetry 1.2 has been released and it broke backward compatibility 🤦, would be great to expedite this.

File: .github/scripts/setup-env.sh [hunk 1]
```diff
@@ -8,7 +8,8 @@
 set -e;
 
 CI_CACHE=$HOME/.cache;
-POETRY_VERSION=1.1.12;
+POETRY_VERSION=1.2.0b3;
```
Comment (enpaul) [inline .github/scripts/setup-env.sh]: We should bump to the 1.12 stable for this PR

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -7,12 +7,11 @@ on:
     branches: ["devel"]
 jobs:
   Test:
+    name: Python ${{ matrix.python.version }}
```
Comment (enpaul) [inline .github/workflows/ci.yaml:10]: :100:

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -39,38 +41,49 @@ jobs:
...
       - name: Configure Job Cache
-        uses: actions/cache@v2
+        uses: actions/cache@v3
         with:
           path: |
             ~/.cache/pip
             ~/.cache/pypoetry/cache
             ~/.poetry
-          # Hardcoded 'py38' slug here lets this cache piggyback on the 'py38' cache
+          # Hardcoded 'py310' slug here lets this cache piggyback on the 'py38' cache
```
Comment (enpaul) [inline .github/workflows/ci.yaml]:
```suggestion
          # Hardcoded 'py310' slug here lets this cache piggyback on the 'py310' cache
```

File: pyproject.toml [hunk 4]
```diff
@@ -35,17 +34,18 @@ classifiers = [
...
+cleo = {version = "^1.0.0a5", optional = true, allow-prereleases = true}
+poetry = {version = "1.2.0b3", optional = true, allow-prereleases = true}
+poetry-core = {version = "1.1.0b3", allow-prereleases = true}
 tox = "^3.8.0"
 
-[tool.poetry.dev-dependencies]
+[tool.poetry.group.dev.dependencies]
 bandit = "^1.6.2"
-black = {version = "^21.12b0", allow-prereleases = true, python = "^3.7"}
+black = "^22.3.0"
```
Comment (enpaul) [inline pyproject.toml:48]: I'd rather move to the stable black in a separate PR rather than rolling it into this change. The reformatting will make this review more error-prone

File: pyproject.toml [hunk 5]
```diff
@@ -35,17 +34,18 @@ classifiers = [
...
+poetry = ["poetry", "cleo"]
 
 [tool.poetry.dependencies]
-python = "^3.6.1"
-poetry = {version = "^1.0.0", optional = true}
-poetry-core = "^1.0.0"
+python = "^3.7"
+cleo = {version = "^1.0.0a5", optional = true, allow-prereleases = true}
+poetry = {version = "1.2.0b3", optional = true, allow-prereleases = true}
+poetry-core = {version = "1.1.0b3", allow-prereleases = true}
```
Comment (enpaul) [inline pyproject.toml]: Can you bump these to the stable versions?

File: pyproject.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "tox-poetry-installer"
-version = "0.8.4"
+version = "1.0a1"
```
Comment (enpaul) [inline pyproject.toml]: I'd rather we bump to `0.9` here rather than 1.0. [The roadmap](https://github.com/enpaul/tox-poetry-installer#path-to-stable) still highlights dropping the `poetry` dependency entirely as a requirement for the 1.0 release and I'd like to keep that the case. Having the plugin require poetry but only as an optional dependency is kind of a gross work around that we need for now until the required functionality gets added to poetry-core. Ideally, once we support poetry 1.2, that can be the next big step.

File: tox_poetry_installer/constants.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 REPORTER_PREFIX: str = f"{__about__.__title__}:"
 
 # Internal list of packages that poetry has deemed unsafe and are excluded from the lockfile
-UNSAFE_PACKAGES: Set[str] = {"distribute", "pip", "setuptools", "wheel"}
+UNSAFE_PACKAGES: Set[str] = set()
```
Comment (enpaul) [inline tox_poetry_installer/constants.py:24]: Can you add a todo comment here noting that the functionality that uses this constant should be removed in a later update, now that it's no longer needed?

File: tox_poetry_installer/utilities.py [hunk 8]
```diff
@@ -182,7 +182,7 @@ def find_project_deps(
     venv: "_poetry.VirtualEnv",
     poetry: "_poetry.Poetry",
     extras: Sequence[str] = (),
-) -> List[PoetryPackage]:
+) -> Set[PoetryPackage]:
```
Comment (enpaul) [inline tox_poetry_installer/utilities.py]: We can't return dependency sequences as sets because python sets are unordered and we need dependencies to be installed in the specified order. The `identify_transients()` function does manual de-dup [here](https://github.com/enpaul/tox-poetry-installer/blob/0.8.5/tox_poetry_installer/utilities.py#L130)

For additional context, #41 is where the problem with using sets was first reported

**Review by enpaul — CHANGES_REQUESTED** [review]
Thanks for doing the legwork here @Callek! There are a few things I highlighted as needing some fixes, but overall it looks good! Really excited to get this merged, and very sorry about the delay!

File: pyproject.toml (hunk 4 above)
Comment (Callek) [reply pyproject.toml:48]: When I don't update black I get:

```
Traceback (most recent call last):
  File "/home/callek/.cache/pypoetry/virtualenvs/tox-poetry-installer-Fk6da_Gk-py3.7/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/callek/.cache/pypoetry/virtualenvs/tox-poetry-installer-Fk6da_Gk-py3.7/lib/python3.7/site-packages/black/__init__.py", line 1372, in patched_main
    patch_click()
  File "/home/callek/.cache/pypoetry/virtualenvs/tox-poetry-installer-Fk6da_Gk-py3.7/lib/python3.7/site-packages/black/__init__.py", line 1358, in patch_click
    from click import _unicodefun
ImportError: cannot import name '_unicodefun' from 'click' (/home/callek/.cache/pypoetry/virtualenvs/tox-poetry-installer-Fk6da_Gk-py3.7/lib/python3.7/site-packages/click/__init__.py)
```

But since I could fix it by updating black I didn't delve too deeply into it

File: pyproject.toml (hunk 5 above)
Comment (Callek) [reply pyproject.toml]: Cleo is still alpha-5 even with poetry 1.2... but the rest :+1:

File: pyproject.toml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "tox-poetry-installer"
-version = "0.8.5"
+version = "0.9"
```
Comment (enpaul) [inline pyproject.toml]:
```suggestion
version = "0.9.0"
```

File: pyproject.toml [hunk 10]
```diff
@@ -35,17 +34,18 @@ classifiers = [
...
+poetry-core = "^1.1.0"
 tox = "^3.8.0"
 
-[tool.poetry.dev-dependencies]
+[tool.poetry.group.dev.dependencies]
 bandit = "^1.6.2"
-black = { version = "^21.12b0", allow-prereleases = true, python = "^3.7" }
+black = "^22.3.0"
 blacken-docs = { version = "^1.8.0", python = "^3.7" }
 ipython = { version = "^7.18.1", python = "^3.7" }
```
Comment (enpaul) [inline pyproject.toml]: Now that the project relies on `^3.7` we can remove the `^3.7` markers from the dev-deps

File: tox_poetry_installer/__about__.py [hunk 11]
```diff
@@ -1,7 +1,7 @@
 # pylint: disable=missing-docstring
 __title__ = "tox-poetry-installer"
 __summary__ = "A plugin for Tox that lets you install test environment dependencies from the Poetry lockfile"
-__version__ = "0.8.5"
+__version__ = "0.9"
```
Comment (enpaul) [inline tox_poetry_installer/__about__.py]:
```suggestion
__version__ = "0.9.0"
```

File: tox_poetry_installer/utilities.py [hunk 12]
```diff
@@ -243,7 +243,7 @@ def find_additional_deps(
             dep_name.lower(), packages, venv, allow_missing=[poetry.package.name]
         )
 
-    return dependencies
+    return dedupe_packages(dependencies)
```
Comment (enpaul) [inline tox_poetry_installer/utilities.py:246]: Rather than calling dedupe in a bunch of places throughout the codebase, could we instead call it just once on the [`packages`](https://github.com/enpaul/tox-poetry-installer/blob/devel/tox_poetry_installer/installer.py#L25) parameter to the `installer.install()` function? I'll admit I haven't done a full eval of all the secondary effects that may have, but that seems both cleaner and more efficient to me, so if it works I'd like to go with that approach instead.

**Review by enpaul — CHANGES_REQUESTED** [review]

File: tox_poetry_installer/utilities.py (hunk 12 above)
Comment (Callek) [reply tox_poetry_installer/utilities.py:246]: There is a bunch of logging/sanity checks along the way that make more sense if we're consistently doing that dedupe: e.g. `Identified {len(dev_deps)} development dependencies to install to env`

File: CHANGELOG.md [hunk 13]
```diff
@@ -2,6 +2,16 @@
...
+## Version 1.0a1
+
+View this release on:
+[Github](https://github.com/enpaul/tox-poetry-installer/releases/tag/1.0a1),
+[PyPI](https://pypi.org/project/tox-poetry-installer/1.0a1/)
+
+- Drop python3.6 support, since Poetry 1.2 dropped py3.6 support
+- Support Poetry==1.2.0b3 (prerelease)
+  - Added direct dep on cleo (prerelease) to support passing in NullIO to poetry functions
+
```
Comment (enpaul) [inline CHANGELOG.md]:
```suggestion
## Version 0.9.0

View this release on:
[Github](https://github.com/enpaul/tox-poetry-installer/releases/tag/0.9.0),
[PyPI](https://pypi.org/project/tox-poetry-installer/0.9.0/)

- Add support for Poetry-1.2.x (#73)
- Update Black formatting to stable release version
- Remove support for Python-3.6
- Remove support for Poetry-1.1.x
- Fix installing dependencies multiple times when transient dependencies are duplicated in the dependency tree
```

**Review by enpaul — CHANGES_REQUESTED** [review]
Last thing :tada:

**Review by enpaul — APPROVED** [review]
